Automattic/vip-cli
# Add dev-env configuration file

![vip-cli-basic-demo](https://user-images.githubusercontent.com/17870752/209755163-c335da2d-065b-49e3-8b02-5299072978d1.gif)

## Description

Adds a configuration file for `vip dev-env` like [`.wp-env.json` for `wp-env`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-env/#wp-env-json). Here's an example configuration file (altered since  the video above):

```yaml
configuration-version: 0.preview-unstable
slug: go-skeleton
title: VIP Go Skeleton
php: 8.0
wordpress: 6.1
xdebug: false
mailhog: false
multisite: false
mu-plugins: image
app-code: ./
elasticsearch: false
phpmyadmin: false
```

This feature is intended to solve a few use-cases:

- When writing documentation for `vip dev-env`, a configuration file can be provided to skip the wizard and reduce setup friction.
- Teams using `vip dev-env` for development can share a source-controlled set of development environment settings
- Makes managing multiple slugs for multiple environments easier

## Design choices

I tried to limit the scope of this PR to provide a useful configuration file, but not one that provides all the features I'd like to see. I've documented those in the next section for future improvement.

[See the discussion below](https://github.com/Automattic/vip-cli/pull/1173#issuecomment-1318083531) for how configuration file options are used. In short, during `vip dev-env create` and `vip dev-env update` the configuration file options from `.vip-dev-env.yml` are merged with command-line parameters and treated the same. If any options are missing, the wizard will prompt for missing options during `vip dev-env start`. If command-line parameters are used with a configuration file, command-line parameters override configuration file options.

For all other `vip dev-env` commands, the `slug` parameter is pulled from the configuration file if present in the current directory, unless a `--slug` is manually specified.

Currently the majority of configuration file functionality is provided in `dev-environment-configuration-file.js`, rather than extending the existing `-cli`/`-core` modules. This was easier for me during development and keeps files smaller, but if desired I'd be happy to move the changes into the existing modules.

## Future feature ideas

This PR is missing some features that might be good for future inclusion:

- Combine `vip dev-env create` and `vip dev-env start`. It would be great if a single command could do both parts when a configuration file is present like `wp-env start`.
- When a new `create` option is added but not yet present in the configuration file, this causes the wizard to prompt during `vip dev-env create`. Ideally, it would be great to have a `useDefaults: true` option or just default to automatically use sane defaults for features not explicitly versioned. This would be helpful for quickly starting up an environment and refining options as needed (also like `wp-env`).
- Provide environment variable configuration via configuration file, like:

    ```yaml
    slug: my-site
    env:
      SOME_VAR: some-value
    ```
- Similar to above, provide `wp-config.php` constant values like [`wp-env.json`'s `config` option](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-env/#wp-env-json).
    ```yaml
    slug: my-site
    config:
      WP_SCRIPT_DEBUG: true
    ```
- In `vip dev-env update`, show specifically which configuration values have changed.
- (maybe) VIP could auto-generate these files for a customer's environment to make initial setup easy.
- (maybe) Search for configuration file in all parent directories like `npm` and `docker-compose`. The current version only searches the current directory.

## Steps to Test

These changes can be tested globally via `npm run build && npm link` using the `add/configuration-file` branch, or commands can be run directly from compiled files like this:

```bash
$ cd vip-cli
$ npm run build
$ node dist/bin/vip-dev-env-start.js
```

### `vip dev-env create`

1. In an existing VIP Go site source or using [vip-go-skeleton](https://github.com/Automattic/vip-go-skeleton), create a `.vip-dev-env.yml` file. Add a top-level slug option and at least one of the following options:

    ```yaml
    configuration-version: 0.preview-unstable
    slug: test-slug
    title: VIP Go Skeleton
    php: 8.0
    wordpress: 6.1
    xdebug: false
    mailhog: false
    multisite: false
    mu-plugins: image  # "image" or path to local mu-plugins/ directory
    app-code: ./       # "image" or path to local app code
    elasticsearch: false
    phpmyadmin: false
    ```

2. Run `vip dev-env create`. This should show all configuration from `.vip-dev-env.yml` that passes validation:

    ```bash
    $ vip dev-env create
    Using environment go-skeleton from .vip-dev-env.yml

    # ...

    Using configuration from file.
    configuration-version: 0.preview-unstable
    slug: wp-content-hub
    title: Content Hub
    multisite: false
    php: 8.0
    wordpress: 6.1
    mu-plugins: image
    app-code: ./
    elasticsearch: false
    phpmyadmin: false
    xdebug: false
    mailhog: false
    ```

3. Run `vip dev-env destroy`, and try removing some values from the configuration file.
4. Run `vip dev-env create` again, and confirm missing options are prompted by the wizard.
5. Run `vip dev-env start`, and ensure options (e.g. `elasticsearch: true`) are respected in the resulting provisioning.

### `vip dev-env update`

If changes are made to the configuration file after a `vip dev-env start`, these will not be propegated in the instance configuration until a `vip dev-env update` is run. This works exactly like passing command-line parameters to `vip dev-env update` - only options specified in the configuration file will be updated to match. If command-line parameters are also passed when a configuration file is present, those parameters will override the configuration file. To test:

1. Create a `.vip-dev-env.yml` file with configuration `phpmyadmin: false` in a VIP Go site source folder. See step 2 above for an example configuration file.
2. Run `vip dev-env create` and `vip dev-env start` to start the environment.
3. View the contents of the local `instance_data.json` file for the created environment:

    ```bash
    $ cat ~/.local/share/vip/dev-environment/<site-slug>/instance_data.json
    {"wpTitle":"VIP Go Skeleton","multisite":false,"elasticsearch":false,"php":"ghcr.io/automattic/vip-container-images/php-fpm:8.0","mariadb":"10.3","mediaRedirectDomain":"","wordpress":{"mode":"image","tag":"6.1"},"muPlugins":{"mode":"image","tag":"image"},"appCode":{"mode":"image","tag":"."},"statsd":false,"phpmyadmin":false,"xdebug":false,"xdebugConfig":"","siteSlug":"go-skeleton","mailhog":false}
    ```

    See the `"phpmyadmin"` key is set to `false`.

4. Edit `.vip-dev-env.yml` and turn on PHPMyAdmin with `phpmyadmin: true`.
5. Run `vip dev-env update`.
6. Reread instance data in `instance_data.json`:

    ```bash
    $ cat ~/.local/share/vip/dev-environment/<site-slug>/instance_data.json
    ```

    Ensure the `"phpmyadmin"` key is now set to `true`.

7. Run `vip dev-env start` to recreate the environment. Ensure that a `phpmyadmin` container is started and the info table contains working `PHPMYADMIN URLS`.

### Other commands

All other `vip dev-env` commands that require a slug (`vip dev-env info`, `vip dev-env exec`) will now pull the slug from a local configuration file if present via `getEnvironmentName()`. Ensure commands like `vip dev-env exec -- wp option get siteurl` work without a slug. If a slug is explicitly provided via command-line parameter, the configuration file slug should be ignored.
